mavros/src/plugins/rc_io.cpp

```diff
@@ -95,13 +96,49 @@ class RCIOPlugin : public plugin::PluginBase {
 		raw_rc_in[offset + 7] = port.chan8_raw;
 		// [[[end]]] (checksum: fcb14b1ddfff9ce7dd02f5bd03825cff)
 
-		auto rcin_msg = boost::make_shared<mavros_msgs::RCIn>();
+		send_rc_data(m_uas->synchronise_stamp(port.time_boot_ms), port.rssi);
+	}
+
+	void send_rc_data(const ros::Time& stamp, const int rssi){
 
-		rcin_msg->header.stamp = m_uas->synchronise_stamp(port.time_boot_ms);
-		rcin_msg->rssi = port.rssi;
-		rcin_msg->channels = raw_rc_in;
+		constexpr double kStickMax = 2000;
+		constexpr double kStickMin = 1000;
 
-		rc_in_pub.publish(rcin_msg);
+		sensor_msgs::Joy rcin_msg;
+ 		rcin_msg.header.stamp = stamp;
+		//if rssi > 0, rc is on and so we set first button
+		rcin_msg.buttons.push_back(rssi > 0);
+ 		if(raw_rc_in.size() < 6){
+			ROS_FATAL("RC has too few control inputs and cannot be used, DO NOT FLY");
+		}
+		else{
+ 			std::vector<double> inputs;
+			for(const uint16_t& raw_rc_channel : raw_rc_in){
+			//convert to range -1 to 1
+				inputs.push_back((static_cast<double>(raw_rc_channel) - kStickMin) / (kStickMax - kStickMin));
+				inputs.back() = 2*inputs.back() - 1;
+				//ensure it is really in range
+				inputs.back() = std::min(std::max(inputs.back(),-1.0),1.0);
+			}
+ 			//force switches to be in extreme positions
+			for(size_t i = 4; i < 7; ++i){
+				//1 is active so we are a bit conservative here
+				if(inputs[i] < 0.5){
+					inputs[i] = -1.0;
+				}else{
+					inputs[i] = 1.0;
+				}
+			}
+ 			//weird ordering and inversions to match asctec interface
```

Comment:
Lol. This explains quite a bit. I remember we spent about a day with helen trying to figure out the correct rc layout for the mpc to work when we first build penguin, because it was doing weird stuff. But this had no effect on manual mode i think. The issue there was that in the settings of the px4 the imu was rotated by 180 degrees and therefore the platform was doing exactly the opposite of what you wanted it to do